### PR TITLE
Add B-Lora training option to the advanced dreambooth lora script

### PR DESCRIPTION
**Pull request by linoytsaban**
Adds an option to train a dreambooth lora targeting using a single ref image + targeting specific unet blocks only,  subsequently allowing for easy combination of style/content LoRAs,
as proposed in [Implicit Style-Content Separation using B-LoRA](https://b-lora.github.io/B-LoRA/)

### CODE REVIEW

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 1]
```diff
@@ -73,9 +74,14 @@
     convert_unet_state_dict_to_peft,
     is_wandb_available,
 )
+from diffusers.utils.hub_utils import load_or_create_model_card, populate_model_card
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:77]: Going forward, prefer not to introduce multiple unrelated changes in a single PR. For example, this PR should primarily about B-LoRA training but here we're also refactoring the model card population related stuff.

**Review by sayakpaul — COMMENTED** [review]
Looking nice. How are the results?

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py (hunk 1 above)
Comment (linoytsaban) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:77]: oh yeah that's actually not meant to go here, will remove that :)

**Comment by linoytsaban** [comment]
some comparisons with InstantStyle: 
Group 1-13

while B-LoRA requires optimization, it can have more robust results depending on the use case+ it requires very light optimization (i.e. I used the default settings for all experiments which consist of a single training image, and 1000 training steps)

will also add examples of multi-concept B-LoRAs here shortly :)

File: examples/advanced_diffusion_training/README.md [hunk 2]
```diff
@@ -304,6 +304,60 @@ accelerate launch train_dreambooth_lora_sdxl_advanced.py \
...
+
+style_B_lora_path = "./"
+content_B_lora_path = "./"
+
+content_B_LoRA = load_blora(content_B_lora_path,"content",1)
+style_B_LoRA = load_blora(style_B_lora_path,"style",1.1)
+combined_lora = {**content_B_LoRA, **style_B_LoRA}
+
+# Load both loras
+pipeline.load_lora_into_unet(res_lora, None, pipeline.unet)
```
Comment (apolinario) [inline examples/advanced_diffusion_training/README.md:355]: `res_lora` isn't anywhere. Should it be `combined_lora` here?

Comment (linoytsaban) [reply examples/advanced_diffusion_training/README.md:355]: yes! I'll fix that + add paths to the LoRAs

File: examples/advanced_diffusion_training/README.md [hunk 3]
```diff
@@ -304,6 +304,60 @@ accelerate launch train_dreambooth_lora_sdxl_advanced.py \
...
+### B-LoRA training 
+The advanced script now supports B-LoRA training too!
+> Proposed in [Implicit Style-Content Separation using B-LoRA](https://b-lora.github.io/B-LoRA/static/source/B-LoRA.pdf), 
+B-LoRA is a method that leverages LoRA to implicitly separate the style and content components of a **single** image.
+It was shown that learning the LoRA weights of two specific blocks (referred to as B-LoRAs) 
+achieves style-content separation that cannot be achieved by training each B-LoRA independently. 
+ Once trained, the two B-LoRAs can be used as independent components to allow various image stylization tasks
+
+**Usage**
+Enable B-LoRA training by adding this flag
```
[the remark this answers is not in the document]
Comment (linoytsaban) [reply examples/advanced_diffusion_training/README.md:316]: yes, I'll add these ad mention the single image

```
!accelerate launch train_dreambooth_b-lora_sdxl.py \
 --pretrained_model_name_or_path="stabilityai/stable-diffusion-xl-base-1.0" \
 --instance_data_dir="./style" \
 --output_dir="B-LoRA_chalk" \
 --instance_prompt="a [v28]" \
 --resolution=1024 \
 --rank=64 \
 --train_batch_size=1 \
 --learning_rate=5e-5 \
 --lr_scheduler="constant" \
 --lr_warmup_steps=0 \
 --max_train_steps=1000 \
 --checkpointing_steps=2000 \
 --seed="0" \
 --gradient_checkpointing \
 --mixed_precision="fp16"
 ```

**Comment by linoytsaban** [comment]
@asomoza 
Nice! we could maybe follow up on some of the experiments/features we covered here:  https://huggingface.co/blog/sdxl_lora_advanced_script

I think that could be a cool avenue to explore, and targeting specific blocks using the current script can be made possible quite easily (basically same as here but as you said not restricted) also for re-training

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 4]
```diff
@@ -77,6 +77,9 @@
 from diffusers.utils.torch_utils import is_compiled_module
 
 
+if is_wandb_available():
+    pass
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:81]: Can be removed I guess?

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 5]
```diff
@@ -740,6 +749,40 @@ def parse_args(input_args=None):
     return args
 
 
+# Taken from B-LoRA repo https://github.com/yardenfren1996/B-LoRA/blob/main/blora_utils.py
+BLOCKS = {
+    "content": ["unet.up_blocks.0.attentions.0"],
+    "style": ["unet.up_blocks.0.attentions.1"],
+}
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:756]: Should we make this configurable instead? 

`content_b_lora_blocks` and `style_b_lora_blocks`?

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 6]
```diff
@@ -1908,12 +1979,6 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
 
                 # Sample noise that we'll add to the latents
                 noise = torch.randn_like(model_input)
-                if args.noise_offset:
-                    # https://www.crosslabs.org//blog/diffusion-with-offset-noise
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1912]: Why are we removing this?

**Review by sayakpaul — APPROVED** [review]
Just minor comments. I think we should add [this](https://github.com/huggingface/diffusers/pull/7741#issuecomment-2074880311) analysis in the README too. I would also love a nice successor blog post (successor to [LoRAs of the world, unite](https://huggingface.co/blog/sdxl_lora_advanced_script)) on B-LoRA containing some fancy results.

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py (hunk 6 above)
Comment (linoytsaban) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1912]: I just realize it doesn't really mean much for _sdxl_ Loras, but could be discussed in a separate PR as these are unrelated

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py (hunk 5 above)
Comment (linoytsaban) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py]: I wonder if we should already make this a single `--lora_unet_blocks` (or smth similar) to allow for full flexability on the targeted blocks for folks who want to experiment (re: @asomoza comments too), but have the default be the 2 blora blocks?

File: examples/advanced_diffusion_training/README.md [hunk 7]
```diff
@@ -304,6 +304,82 @@ accelerate launch train_dreambooth_lora_sdxl_advanced.py \
...
+The inference is a bit different:
+1. we need load *specific* unet layers (as opposed to a regular LoRA/DoRA)
+2. the trained layers we load, changes based on our objective (e.g. style/content)
+
+```bash
+git clone https://github.com/yardenfren1996/B-LoRA.git
+%cd B-LoRA
+```
+```python
+from blora_utils import BLOCKS, filter_lora, scale_lora
```
Comment (bghira) [inline examples/advanced_diffusion_training/README.md:349]: relying on third party untrusted repo not even pinned to a given commit or release tag is super dangerous

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 8]
```diff
@@ -946,16 +981,20 @@ def __init__(
                 transforms.Normalize([0.5], [0.5]),
             ]
         )
+        # if using B-LoRA for single image. do not use transformations
+        single_image = len(self.instance_images) < 2
         for image in self.instance_images:
-            image = exif_transpose(image)
+            if not single_image:
+                image = exif_transpose(image)
```
Comment (bghira) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:988]: we NEED to do the EXIF transform.

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 9]
```diff
@@ -946,16 +981,20 @@ def __init__(
...
             if not image.mode == "RGB":
                 image = image.convert("RGB")
             self.original_sizes.append((image.height, image.width))
             image = train_resize(image)
-            if args.random_flip and random.random() < 0.5:
+
+            if not single_image and args.random_flip and random.random() < 0.5:
```
Comment (bghira) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:994]: probably better to use the `use_blora` arg instead of disabling this for every training run on a single img

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 10]
```diff
@@ -1229,10 +1268,12 @@ def main(args):
...
 
-        model_id = args.hub_model_id or Path(args.output_dir).name
-        repo_id = None
         if args.push_to_hub:
-            repo_id = create_repo(repo_id=model_id, exist_ok=True, token=args.hub_token).repo_id
+            repo_id = create_repo(
+                repo_id=args.hub_model_id or Path(args.output_dir).name, exist_ok=True, token=args.hub_token
+            ).repo_id
+        else:
+            repo_id = Path(args.output_dir).name
```
Comment (bghira) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1276]: why is this refactored? it was better before, did anything functionally change?

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 11]
```diff
@@ -1374,12 +1422,19 @@ def main(args):
...
+        target_modules = get_blora_target_modules(unet)
+    else:
+        target_modules = ["to_k", "to_q", "to_v", "to_out.0"]
+
     unet_lora_config = LoraConfig(
         r=args.rank,
-        lora_alpha=args.rank,
         use_dora=args.use_dora,
+        lora_alpha=args.rank,
         init_lora_weights="gaussian",
```
Comment (bghira) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1436]: why are we using gaussian unconditionally? i found that `default` performs better and has fewer artifacts earlier on.

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 12]
```diff
@@ -1505,7 +1561,8 @@ def load_model_hook(models, input_dir):
             models = [unet_]
             if args.train_text_encoder:
                 models.extend([text_encoder_one_, text_encoder_two_])
-            cast_training_params(models)
+                # only upcast trainable parameters (LoRA) into fp32
+                cast_training_params(models)
```
Comment (bghira) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1565]: unless you're using the AnyPrecision optimiser or the adamw-bf16 optimiser this shouldn't be indented. and in fact for AnyPrecision or AdamW-BF16 we should not run this at all. it should run whether or not the TE is trained.

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 13]
```diff
@@ -1525,6 +1582,8 @@ def load_model_hook(models, input_dir):
         models = [unet]
         if args.train_text_encoder:
             models.extend([text_encoder_one, text_encoder_two])
+
+        # only upcast trainable parameters (LoRA) into fp32
         cast_training_params(models, dtype=torch.float32)
```
Comment (bghira) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1587]: actually, this happens twice? this time it happens outside of teh TE block, which is correct

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py (hunk 6 above)
Comment (bghira) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1912]: this cannot be removed @linoytsaban

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 14]
```diff
@@ -2139,10 +2197,6 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
 
         if accelerator.is_main_process:
             if args.validation_prompt is not None and epoch % args.validation_epochs == 0:
-                logger.info(
-                    f"Running validation... \n Generating {args.num_validation_images} images with prompt:"
-                    f" {args.validation_prompt}."
-                )
```
Comment (bghira) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:2145]: why remove this?

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py (hunk 5 above)
Comment (bghira) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py]: if making it configurable, you'll have to check the values and error out when they don't exist. it would be nice if that also shows the possible correct values to use.

File: examples/advanced_diffusion_training/README.md (hunk 7 above)
Comment (apolinario) [reply examples/advanced_diffusion_training/README.md:349]: I agree that maybe pinning to a current commit would make sense here

**Review by apolinario — APPROVED** [review]
LGTM

**Comment by bghira** [comment]
thanks for the cool feature @linoytsaban did you try anything with a few images or a large set to see if it scales at all? or does it require a single img for this to function?

File: examples/advanced_diffusion_training/README.md (hunk 7 above)
Comment (sayakpaul) [reply examples/advanced_diffusion_training/README.md:349]: @linoytsaban I think this is yet to be addressed.

File: examples/advanced_diffusion_training/README.md [hunk 15]
```diff
@@ -304,6 +304,83 @@ accelerate launch train_dreambooth_lora_sdxl_advanced.py \
...
+  B_LoRA = filter_lora(B_LoRA_sd, BLOCKS[type])
+  B_LoRA = scale_lora(B_LoRA, alpha)
+  return B_LoRA
+
+vae = AutoencoderKL.from_pretrained("madebyollin/sdxl-vae-fp16-fix", torch_dtype=torch.float16)
+pipeline = StableDiffusionXLPipeline.from_pretrained(
+        "stabilityai/stable-diffusion-xl-base-1.0",
+        vae=vae,
+        torch_dtype=torch.float16,
+    ).to("cuda")
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/README.md:364]:
```suggestion
pipeline = StableDiffusionXLPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0",
    vae=vae,
    torch_dtype=torch.float16,
).to("cuda")
```

**Review by sayakpaul — APPROVED** [review]
Just two small nits. After that let's ship this

**Comment by linoytsaban** [comment]
Made a few changes to readme and inference re: making lora unet blocks generally configurable 
If you do mind taking a final look @apolinario @sayakpaul @asomoza , and then 🛳️

File: examples/advanced_diffusion_training/README.md [hunk 16]
```diff
@@ -304,6 +304,147 @@ accelerate launch train_dreambooth_lora_sdxl_advanced.py \
 > [!CAUTION]
 > Min-SNR gamma is not supported with the EDM-style training yet. When training with the PlaygroundAI model, it's recommended to not pass any "variant".
 
+### B-LoRA training 
+The advanced script now supports B-LoRA training too!
+> Proposed in [Implicit Style-Content Separation using B-LoRA](https://b-lora.github.io/B-LoRA/static/source/B-LoRA.pdf),
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/README.md:309]:
```suggestion
> Proposed in [Implicit Style-Content Separation using B-LoRA](https://arxiv.org/abs/2403.14572), 
```

File: examples/advanced_diffusion_training/README.md [hunk 17]
```diff
@@ -304,6 +304,147 @@ accelerate launch train_dreambooth_lora_sdxl_advanced.py \
...
+B-LoRA is a method that leverages LoRA to implicitly separate the style and content components of a **single** image.
+It was shown that learning the LoRA weights of two specific blocks (referred to as B-LoRAs) 
+achieves style-content separation that cannot be achieved by training each B-LoRA independently. 
+Once trained, the two B-LoRAs can be used as independent components to allow various image stylization tasks
+
+**Usage**
+Enable B-LoRA training by adding this flag
+```bash
+--use_blora
+```
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/README.md:319]:
```suggestion
```diff
+ --use_blora
```
```

**Comment by sayakpaul** [comment]
LGTM! Thanks for eliminating the dependency altogether. 

Will merge once the CI is green pass.

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py (hunk 8 above)
Comment (bghira) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1009]: i don't get why the single image isn't exif transformed still, this seems like a bug
